商品出品機能

nana0817 opened the pull request:
# What
Furimaアプリに商品出品機能を実装
# Why
Furimaアプリで商品を出品できるようにするため。
### ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/7565372d2c00f4c8467d32981af20d5a
### 価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/80785c06f92905bb767cabcdc9195b26
### 必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/88950a7cdf12628289809140026b6372
### 入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/0a6324e43a43b74e58571dd68a4fb3f1
### ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/c065826e3605eb412ee8cddfc612fa8d
### テスト結果の画像
https://i.gyazo.com/b3a912ae588ac06f68d9eec8fa845fbb.png

nana0817 committed:
Itemモデル、itemsテーブルの作成

rails g model itemでモデルを作成し、
同時生成されたマイグレーションファイルにカラム名等の設定を記述した後に、マイグレートしてテーブルを作成した。

nana0817 committed:
モデルファイルにアソシエーションを記述

nana0817 committed:
Itemモデルのバリデーションを記述

nana0817 committed:
itemsコントローラーのnew, createアクションのルーティングを追加

nana0817 committed:
app/views/items/index.html.erbファイルにnewアクションのパスを設定

nana0817 committed:
未ログインユーザーのリダイレクト処理(newアクション)

未ログインユーザーが商品出品ページにアクセスしようとした時、ログインページへリダイレクトするように設定した

nana0817 committed:
app/views/items/new.html.erb編集

・form_withのmodelとurlの属性値を設定
・コメントアウトを外して、エラーメッセージが表示されるようにした。

nana0817 committed:
ActiveHashの導入

以下を実行
・必要なGem(active_hash)をインストール
・app/modelsディレクトリに「category.rb」「condiion.rb」「shipping_fee.rb」「prefecture.rb」「shipping_time.rb」ファイルを作成
・作成したファイルに必要な記述を追加
・作成したファイルそれぞれにItemモデルとのアソシエーションを記述
・app/models/item.rbファイルに作成したファイルとのアソシエーションを記述
・app/models/item.rbファイルにバリデーションを追加
・app/views/items/new.html.erbファイルにActiveHashのための記述を追加
・app/views/items/new.html.erbファイルのname属性がhogeとなっていたところを適切な値に変更

nana0817 committed:
画像アップロード機能の実装

以下を実行
・実装に必要なGemをインストール
・画像処理にMiniMagickを使用するため、config/application.rbにて設定を変更
・Active Storageをインストールし、生成されたマイグレーションファイルをマイグレート
・Active StorageのテーブルとMessagesテーブルのアソシエーションを定義
・ストロングパラメーターを定義
・imageのバリデーションを追加

nana0817 committed:
出品に失敗した際のエラーハンドリング実装

以下を実行
・app/controllers/items_controller.rbのcreateアクションを編集
・app/modelsディレクトリのshipping_fee.rbファイルとshipping_time.rbファイルでモデルクラスの命名に問題があったので修正。加えて、app/views/items/new.html.erbファイルの記述も修整。
・app/views/items/new.html.erbファイルでエラーメッセージを表示する記述のERBタグのイコールが抜けていたため、修正。
・priceのバリデーションに不備があったので、修正

nana0817 committed:
入力した価格によって販売手数料と販売利益の表示が変わる仕組みを実装

以下を実行
・javasciptのファイルを作成
・作成したファイルを読み込めるよう設定を追加
・作成したファイルにコードを追加

nana0817 committed:
テストコード実装

以下を実行
・app/models/item.rbに文字数制限のバリデーションを追加
・app/views/items/new.html.erbの文字数制限の記述を削除
・テストファイルを作成
・FactoryBotでダミーデータを作成する記述を追加
・テストコードを記述

nana0817 committed:
Rubocopでコードを自動修正

CHURO-MAU commented on app/controllers/application_controller.rb:
```diff
@@ -1,6 +1,7 @@
 class ApplicationController < ActionController::Base
   before_action :basic_auth
   before_action :configure_permitted_parameters, if: :devise_controller?
+  before_action :authenticate_user!, only: [:new]
```
４行目
authenticate_user!メソッドは適切なコントローラーへ記述しましょう。

application_controller.rbの場合、すべての処理で:authenticate_user!メソッドが読み込まれてしまいます。現状ですと、全てのコントローラーのnewアクションに適用されている状態です。

:authenticate_user!メソッドを適用したい処理のみで読み込まれるように、適切なコントローラーへ記述しましょう。

CHURO-MAU commented on app/models/item.rb:
```diff
@@ -0,0 +1,35 @@
+    validates :shipping_time_id  # 発送までの日数
+  end
+
+  validates :price, presence: true, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999,
```
３３行目
整数のみを許可するようにバリデーションを修正してみましょう。

現状ですと、少数値でも許可されてしまう状態です。

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！

コメントいたしましたので、ご確認お願いいたします。
修正後は、再度レビュー依頼フォームのご提出をお願いいたします。

nana0817 commented:
コードレビューありがとうございます。
早速修正に取り掛かります！

nana0817 committed:
authenticate_user!メソッドを適切なコントローラーへ記述

前の状態だと、itemsコントローラー以外のnewアクションにもauthenticate_user!メソッドが適用される状態になっていた。
今のところ、authenticate_user!が適用されるのはitems#newだけでいいので、記述をitemsコントローラーへ移動。

nana0817 committed:
priceのバリデーションを整数のみ許可するよう修正

前の状態だと、小数値でも許可されてしまう状態だったので、バリデーションを修正。
また、numericalityがあれば、presence: trueはいらないので削除。

nana0817 committed:
バリデーション。テストコード　修正＆Rubocopで自動修正

nana0817 committed:
バリデーション　謎の改行を修正

nana0817 commented:
ご指摘いただいた点は全て修正しました！

最新のテストコードの結果です。
https://i.gyazo.com/1aaca2bdb077b89e428ce5963caff856.png

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます！実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう！

nana0817 committed:
バリデーション・テストコードを修正、Rubocopで自動修正
